rxjava-1-examples/src/main/java/io/vertx/example/rxjava/database/jdbc/Transaction.java

```diff
@@ -0,0 +1,55 @@
+package io.vertx.example.rxjava.database.jdbc;
+
+import io.vertx.core.json.JsonArray;
+import io.vertx.core.json.JsonObject;
+import io.vertx.example.util.Runner;
+import io.vertx.rxjava.core.AbstractVerticle;
+import io.vertx.rxjava.ext.jdbc.JDBCClient;
+
+/*
+ * @author <a href="mailto:emad.albloushi@gmail.com">Emad Alblueshi</a>
+ */
+public class Transaction extends AbstractVerticle {
+
+  // Convenience method so you can run it in your IDE
+  public static void main(String[] args) {
+    Runner.runExample(Transaction.class);
+  }
+
+  @Override
+  public void start() throws Exception {
+
+    JsonObject config = new JsonObject().put("url", "jdbc:hsqldb:mem:test?shutdown=true")
+      .put("driver_class", "org.hsqldb.jdbcDriver");
+
+    String sql = "CREATE TABLE colors (" +
+      "id INTEGER GENERATED BY DEFAULT AS IDENTITY(START WITH 1, INCREMENT BY 1) PRIMARY KEY, " +
+      "name VARCHAR(255), " +
+      "datetime TIMESTAMP WITH TIME ZONE DEFAULT CURRENT_TIMESTAMP NOT NULL)";
+
+    JDBCClient client = JDBCClient.createShared(vertx, config);
+
+    // Connect to the database
+    client
+      .rxGetConnection()
+      .flatMap(conn ->
+        conn
+          // disable auto commit to manage transaction manually
+          .rxSetAutoCommit(false)
+          .flatMap(autoCommit -> conn.rxExecute(sql))
+          .flatMap(executed -> conn.rxUpdateWithParams("INSERT INTO colors (name) VALUES (?)", new JsonArray().add("BLACK")))
+          .flatMap(updateResult -> conn.rxUpdateWithParams("INSERT INTO colors (name) VALUES (?)", new JsonArray().add("WHITE")))
+          .flatMap(updateResult -> conn.rxUpdateWithParams("INSERT INTO colors (name) VALUES (?)", new JsonArray().add("PURPLE")))
+          .flatMap(updateResult -> conn.rxQuery("SELECT * FROM colors"))
+          // commit if all succeeded
+          .doOnSuccess(resultSet -> conn.rxCommit().subscribe())
```

Comment:
In general, I wouldn't recommend nesting subscribes like that. By nesting the subscribes, any error emitted by commit/rollback won't be propagated to the caller. The commit/rollback will also happen asynchronously, which creates a race condition where code downstream can execute before the commit/rollback completes. The semantics also change, since the rollback method won't be called if the commit call fails, but it would have been if you used the flatMap/map combo (and it would have been if the doOnSuccess/doOnError combo worked the way you thought it did).

I guess it might be ok for the example, but I have found it's usually not a good pattern, unless you have considered all the side-effects (no error propagation, stuff downstream can run concurrently) and are ok with them.

This also applies for the RxJava 2 example.

Reply:
@dano Do you mean something like below ?

```javascript
// commit if all succeeded
.flatMap(updateResult -> conn.rxCommit().map(commit -> updateResult))
// rollback if any failed
.onErrorResumeNext(throwable -> conn.rxRollback().flatMap(rollback -> Single.error(throwable)))
```
I agree that commit/rollback method won't be propagated to the caller but in that case which I think a very rare scenario.

If the end user is sicked to the threading model in Vert.x will be he/she safe regarding concurrency ?

Does it make sense if the commit "itself" fails the rollback should be called ?

Reply:
@EmadAlblueshi Yes, though for the rollback case, if rollback were to fail, you'll lose the original error that caused you to try rollback to begin with. To handle that, you need to do something like this:

```java
.onErrorResumeNext(e -> conn.rxRollback()
   .onErrorResumeNext(e2 -> Single.error(new CompositeException(e, e2)))
   .flatMap(ign -> Single.error(e))
)
```
`CompositeException` is an RxJava-provided type used to capture multiple errors in the chain.

> I agree that commit/rollback method won't be propagated to the caller but in that case which I think a very rare scenario.

Well, it depends. If someone is using deferred constraint checking in their transaction, it might not be that rare for a commit to fail. Even if it was rare, you don't ever want your code to hang forever, which is likely what would happen should the commit actually fail.

> If the end user is sticked to the threading model in Vert.x will be he/she safe regarding concurrency?
No. The issue here isn't a shared mutable data problem, it's that stuff that logically is only supposed to run *after* the commit/rollback finishes can now run concurrently to the commit/rollback. So, if right after committing, the next step in the Rx chain was some code queries the DB, and assumes it's going to find whatever you just inserted or modified, it might not find it.

> Does it make sense if the commit "itself" fails the rollback should be called ?
Hmm, probably not in most cases. If the commit failed because you suddenly couldn't connect to the DB, the rollback is probably going to fail, too, so there's no reason to rollback. If the commit failed because a deferred constraint check failed, I think it depends on your specific settings -  I believe for some DBs only the statements with failed constraints are rolled back automatically. In any case, if you don't want the rollback to be called when the commit fails, just move that call to before the commit call in the chain.